perf-regression/gpu-margo-p2p-bw.cu

```diff
@@ -0,0 +1,807 @@
+        ((hg_size_t*)buffer)[i] = i;
+
+    /* Perform cudaMemcpy only if (GPU to GPU || -k option) */
+    if (use_cuda_buf) 
+    	cudaMemcpy(c_buffer, buffer, g_opts.g_buffer_size, cudaMemcpyHostToDevice);
+
+    ret = ssg_get_group_member_addr(gid, target, &target_addr);
+    assert(ret == SSG_SUCCESS);
+
+    ret = margo_create(mid, target_addr, id, &handle);
+    assert(ret == 0);
+
+    if (use_cuda_buf) {
+        bulk_attr.mem_type = (hg_mem_type_t)NA_MEM_TYPE_CUDA;
+	buffer = c_buffer;
+    }
+    /* Use c_buffer if (GPU to GPU or -k option) otherwise use g_buffer */
+    ret = margo_bulk_create_attr(mid, 1, (void **)&buffer, &g_opts.g_buffer_size, 
+			HG_BULK_READWRITE, &bulk_attr, &in.bulk_handle);
+    assert(ret == 0);
+    in.op       = HG_BULK_PULL;
+    in.shutdown = 0;
+    in.duration = duration;
+
+    start_ts = ABT_get_wtime();
+    ret      = margo_forward(handle, &in);
+    end_ts   = ABT_get_wtime();
+    assert(ret == 0);
+
+    ret = margo_get_output(handle, &out);
+    assert(ret == HG_SUCCESS);
+
+    if (print_flag) {
+        printf(
+            "<op>\t<warmup_seconds>\t<concurrency>\t<threads>\t<xfer_size>\t<"
+            "total_bytes>\t<seconds>\t<MiB/"
+            "s>\t<xfer_memory>\t<align_buffer>\n");
+
+        printf("PULL\t%d\t%d\t%d\t%d\t%lu\t%f\t%f\t%lu\t%d\n",
+               g_opts.warmup_seconds, g_opts.concurrency, g_opts.threads,
+               g_opts.xfer_size, out.bytes_moved, (end_ts - start_ts),
+               ((double)out.bytes_moved / (end_ts - start_ts))
+                   / (1024.0 * 1024.0),
+               g_opts.g_buffer_size, g_opts.align_buffer);
+    }
+
+    margo_free_output(handle, &out);
+
+    /* pause a moment */
+    margo_thread_sleep(mid, 100);
+
+    in.op       = HG_BULK_PUSH;
+    in.shutdown = shutdown;
+    in.duration = duration;
+
+    start_ts = ABT_get_wtime();
+    ret      = margo_forward(handle, &in);
+    end_ts   = ABT_get_wtime();
+    assert(ret == 0);
+
+    ret = margo_get_output(handle, &out);
+    assert(ret == HG_SUCCESS);
+
+    if (print_flag) {
+        printf("PUSH\t%d\t%d\t%d\t%d\t%lu\t%f\t%f\t%lu\t%d\n",
+               g_opts.warmup_seconds, g_opts.concurrency, g_opts.threads,
+               g_opts.xfer_size, out.bytes_moved, (end_ts - start_ts),
+               ((double)out.bytes_moved / (end_ts - start_ts))
+                   / (1024.0 * 1024.0),
+               g_opts.g_buffer_size, g_opts.align_buffer);
+    }
+
+    /* calculate how many bytes of the buffer have been transferred */
+    bytes_to_check
+        = (g_opts.g_buffer_size / g_opts.xfer_size) * g_opts.xfer_size;
+    if (out.bytes_moved < bytes_to_check) bytes_to_check = out.bytes_moved;
+    /* check fill pattern we got back; should be what we set plus one */
+
+    /* Perform cudaMemcpy only if (GPU to GPU || -j option) */
+    if (use_cuda_buf) 
+    	cudaMemcpy(g_buffer, c_buffer, g_opts.g_buffer_size, cudaMemcpyDeviceToHost);
```

Comment:
I believe this cudaMemcpy block can also be surrounded with `#ifndef NDEBUG`. It looks like g_buffer is only used here to verify the data is correct with an assertion in debug builds.